File: Difficulty: Easy/Unique substrings/unique-substrings.java
//{ Driver Code Starts
//Initial Template for Java


import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        sc.nextLine();
        while(T-->0)
        {
            String S = sc.nextLine();
            String[] S1 = S.split(" ");
            String s = S1[0];
            int k = Integer.parseInt(S1[1]);
            Solution ob = new Solution();
            long  res  = ob.unique_substring_sum(s, k);
            System.out.println(res);
        }
    }
}


// } Driver Code Ends

// User function Template for Java

class Solution {
    public long unique_substring_sum(String s, int k) {
        // If the length of the string is not divisible by k, return -1
        if (s.length() % k != 0)
            return -1;
        
        // Use a Set to store unique substrings of length k
        Set<String> uniqueSubstrings = new HashSet<>();
        String temp = "";

        // Loop through the string to extract substrings of length k
        for (int i = 0; i < s.length(); i++) {
            temp += s.charAt(i);
            // When we have a substring of length k, add it to the set and reset temp
            if (temp.length() == k) {
                uniqueSubstrings.add(temp);
                temp = "";
            }
        }

        // Calculate the sum of the binary values of the unique substrings
        long sum = 0;
        for (String binaryString : uniqueSubstrings) {
            // Convert each binary string to a decimal integer and add to the sum
            sum += Integer.parseInt(binaryString, 2);
        }
        
        return sum;
    }
}

File: Difficulty: Easy/Unique Substrings/unique-substrings.js
//{ Driver Code Starts
// Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => { inputString += inputStdin; });

process.stdin.on("end", (_) => {
    inputString =
        inputString.trim().split("\n").map((string) => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let str = readLine().trim();
        let obj = new Solution();
        let res = obj.unique_substring(str);
        console.log(res);

        console.log("~");
    }
}

// } Driver Code Ends

// User function Template for javascript

/**
 * @param {string} s
 * @return {number}
 */
class Solution {
    // Function to calculate the number of unique substrings in the given string.
    unique_substring(s) {
        // Create a Set to store unique substrings.
        let uniqueSubstrings = new Set();
        
        // Outer loop to select the starting character of the substring.
        for (let startIndex = 0; startIndex < s.length; startIndex++) {
            let currentSubstring = "";
            
            // Inner loop to extend the substring from the start index.
            for (let endIndex = startIndex; endIndex < s.length; endIndex++) {
                // Append the current character to form a substring.
                currentSubstring += s[endIndex];
                
                // Add the formed substring to the Set.
                uniqueSubstrings.add(currentSubstring);
            }
        }
        
        // Return the total count of unique substrings.
        return uniqueSubstrings.size;
    }
}
